scenario/smartsearch.go

```diff
@@ -0,0 +1,350 @@
+	case SearchTextSourceFile:
+		rowFile, err := helpers.NewRowFile(settings.SearchTextFilePath)
+		if err != nil {
+			return err
+		}
+		settings.SearchTextList = rowFile.Rows()
+	default:
+		return errors.Errorf("Unknown SearchTextSource<%d>", settings.SearchTextSource)
+	}
+	return nil
+}
+
+// Validate implements ActionSettings interface
+func (settings SmartSearchSettings) Validate() ([]string, error) {
+	var warnings []string
+	if len(settings.SearchTextList) == 0 {
+		return warnings, errors.New("no searchtext found")
+	}
+	for idx, searchtext := range settings.SearchTextList {
+		searchTerms := parseSearchTerms(searchtext)
+		if len(searchTerms) == 0 {
+			return warnings, errors.Errorf(`no search terms found in searchtext%d<%s> `, idx+1, settings.SearchTextList[idx])
+		}
+	}
+
+	return warnings, nil
+}
+
+func intPtr(i int) *int {
+	return &i
+}
+
+func quote(s string) string {
+	return fmt.Sprintf(`%#v`, s)
+}
+
+func quoteList(strList []string) string {
+	if len(strList) == 0 {
+		return "NONE"
+	}
+	quoted := []string{}
+	for _, v := range strList {
+		quoted = append(quoted, quote(v))
+	}
+	return strings.Join(quoted, ",")
+}
+
+func ternary(cond bool, ifCondTrue string, ifCondFalse string) string {
+	if cond {
+		return ifCondTrue
+	}
+	return ifCondFalse
+
+}
+
+func logSearchSuggestionResult(logEntry *logger.LogEntry, id int, searchSuggestionResult *enigma.SearchSuggestionResult) {
+	if !logEntry.ShouldLogDebug() {
+		return
+	}
+	suggestions := []string{}
+	for _, suggestion := range searchSuggestionResult.Suggestions {
+		suggestions = append(suggestions, suggestion.Value)
+	}
+	logEntry.LogDebugf("search%d suggestions: %s", id, quoteList(suggestions))
+}
+
+func logSearchResult(logEntry *logger.LogEntry, id int, searchResult *enigma.SearchResult) {
+	if !logEntry.ShouldLogDebug() {
+		return
+	}
+	hasResult := false
+	for sgIdx, sga := range searchResult.SearchGroupArray {
+		for _, item := range sga.Items {
+			searchTermsMatched := []string{}
+			for _, stIndex := range item.SearchTermsMatched {
+				searchTermsMatched = append(searchTermsMatched, searchResult.SearchTerms[stIndex])
+			}
+			matches := []string{}
+			for _, match := range item.ItemMatches {
+				matches = append(matches, match.Text)
+			}
+			hasResult = true
+			logEntry.LogDebugf(`search%d result group%d matching term%s %s in %s %s: %s`, id, sgIdx+1,
+				ternary(len(searchTermsMatched) != 1, "s", ""), quoteList(searchTermsMatched),
+				item.ItemType, quote(item.Identifier), quoteList(matches))
+		}
+	}
+	if !hasResult {
+		logEntry.LogDebugf("search%d result: NONE", id)
+	}
+}
+
+var spaceRegex = regexp.MustCompile(`\s+`)
+
+func parseSearchTerms(s string) []string {
+	searchTerms := []string{}
+	sb := &strings.Builder{}
+	quoted := false
+	escaped := false
+	addSearchTerm := func() {
+		defer sb.Reset()
+		searchTerm := sb.String()
+		searchTerm = strings.TrimSpace(searchTerm)
+		searchTerm = spaceRegex.ReplaceAllString(searchTerm, " ")
```

Comment:
regex is usually heavy and slow, if we feel we can't solve it with string split or similar might want to have this done in unmarshal or similar not have every user have to do it for every invoke.

Reply:
Oh yeah, didn't even notice that. Would be better if we could move the entire `parseSearchTerms` into running with Unmarshal. Problem is, that interferes with the chunk pre-processing step since that involves randomization per-user. If you can think of a way to re-architect it such that we don't need to do regex or splitting in the Execute function, that would be a worthwhile improvement.

Reply:
Splitting must be done in the execute function because of randomized typing as @SeBBBe said. The regex on the other hand is not that important, so I'll remove it or replace it with something more efficient. I noticed that if I wrote `"hello     world"` it became one search term `hello world`.

Reply:
Isn't it possible to run parseSearchTerms in the unmarshal and then run some lighter weight logic in the Execute? Since the chunk will always consist of x whole search terms, and possibly ending in a cut off one.